acquia_lift.admin.inc

```diff
@@ -858,3 +858,866 @@ function acquia_lift_report_conversion() {
+  $value_types = array();
+  $operator_options = array(
+    'string' => array(
+      'equals' => t('equals'),
+      'contains' => t('contains'),
+      'starts' => t('starts with'),
+      'ends' => t('ends with')
+    ),
+    'number' => array(
+      'equals' => t('equals'),
+      'numgt' => t('greater than'),
+      'numlt' => t('less than')
+    )
+  );
+
+  foreach ($targeting_values as  $key => $info) {
+    $option_key = $info['visitor_context'] . PERSONALIZE_TARGETING_ADMIN_SEPARATOR . $key;
+    $key_label = isset($info['friendly name']) ? $info['friendly name'] : $key;
+    $context_options[$option_key] = $key_label;
+    $value_types[$option_key] = $info['value type'];
+  }
+
+  // This is the portion of the form that will be replace when the "add new" or "remove
+  // context" links are clicked.
+  $main_wrapper_id = 'acquia-lift-targeting';
+
+  $form['mapping'] = array(
+    '#tree' => TRUE,
+    '#theme_wrappers' => array('container'),
+    '#attributes' => $attributes,
+  );
+  $form['mapping']['contexts'] = array(
+    '#type' => 'container',
+    '#attributes' => array(
+      'id' => $main_wrapper_id,
+    ),
+  );
+
+  // Load from the existing form if passed first, otherwise, load from option
+  // set.
+  $mappings = array();
+  if (isset($form_state['values']['mapping']['contexts'])) {
+    foreach($form_state['values']['mapping']['contexts'] as $delta => $context) {
+      list($plugin_name, $context_option) = explode(PERSONALIZE_TARGETING_ADMIN_SEPARATOR, $context['context']);
+      // Important: preserve the delta passed through the form as it is used
+      // to determine the item to delete when "remove" is clicked.
+      $mappings[$delta] = array(
+        'plugin' => $plugin_name,
+        'context' => $context_option,
+        'operator' => $context['value']['operator'],
+        'match' => $context['value']['match'],
+      );
+    }
+  }
+
+  // If the "Remove" button was clicked for a context, we need to remove that context
+  // from the form.
+  if (isset($form_state['to_remove'])) {
+    unset($mappings[$form_state['to_remove']]);
+    unset($form_state['to_remove']);
+    $form_state['num_contexts']--;
+  }
+
+  // Make sure there is at least an empty context.
+  if (empty($mappings)) {
+    $mappings[] = array(
+      'context' => '',
+      'operator' => 'equals',
+      'match' => '',
+      'plugin' => '',
+    );
+  }
+
+  // If the "Add another" button was clicked, we need to add contexts to get up
+  // to the number indicated.
+  $num_contexts = count($mappings);
+  if (isset($form_state['num_contexts']) && $form_state['num_contexts'] > $num_contexts) {
+    while ($num_contexts < $form_state['num_contexts']) {
+      $mappings[] = array(
+        'context' => '',
+        'operator' => 'equals',
+        'match' => '',
+        'plugin' => '',
+      );
+      $num_contexts++;
+    }
+  }
+
+  $form_state['num_contexts'] = count($mappings);
+  foreach ($mappings as $delta => $mapping) {
```

Comment:
Is all this logic duplicated?  We're going to continue to have it in both places... although I guess maybe I'll split that out once I work with it because I'll know what portions I'm re-using.